### PR TITLE
Add torch.take and Tensor.put_

### PR DESCRIPTION
**Pull request by colesbury**
These are similar to numpy.take and numpy.put. The take function allows
you to linearly index into a tensor without viewing it as a 1D tensor
first. The output has the same shape as the indices. The put function
copies value into a tensor also using linear indices.

I'll add these to ATen+Variable in a subsequent PR

### CODE REVIEW

**Comment by colesbury** [comment]
Two differences from the NumPy implementations:

https://docs.scipy.org/doc/numpy-1.13.0/reference/generated/numpy.take.html
https://docs.scipy.org/doc/numpy-1.13.0/reference/generated/numpy.put.html

`numpy.take` has an optional axis argument, which behaves like `index_select`. I didn't include this. We can add it later without breaking backwards compatibility.

`numpy.put` repeats the values if necessary to make them as long as indices. This is strange, since most other functions use broadcasting semantics. Currently, `Tensor.put_` requires values to have the same length as indices. We can change this without breaking backwards compatibility.

File: torch/lib/TH/generic/THTensorMath.c [hunk 1]
```diff
@@ -349,6 +349,81 @@ void THTensor_(indexCopy)(THTensor *tensor, int dim, THLongTensor *index, THTens
...
+
+static int64_t THTensor_(wrapLinearIndex)(int64_t linearIndex, int64_t numel) {
+  THArgCheck(linearIndex < numel && linearIndex >= -numel, 2, "out of range");
+  return linearIndex < 0 ? linearIndex + numel : linearIndex;
+}
+
+void THTensor_(take)(THTensor *r_, THTensor *src, THLongTensor *index)
+{
+  THTensor_(resizeNd)(r_, index->nDimension, index->size, NULL);
+  THTensor* dst = THTensor_(newContiguous)(r_);
```
Comment (apaszke) [inline torch/lib/TH/generic/THTensorMath.c]: Doesn't `resizeNd` ensure that `r_` is contiguous?

File: torch/lib/THC/THCTensorIndex.cu [hunk 2]
```diff
@@ -398,5 +398,53 @@ __global__ void calculateLinearIndices(
...
 }
 
+template <typename T, typename IndexType, int Dims>
+struct TensorPutOp {
+  TensorPutOp(TensorInfo<T, IndexType> info) : info(info) {}
+
+  __device__ __forceinline__ void operator()(T* value, int64_t* index) {
+    IndexType linearIndex = static_cast<IndexType>(*index);
+    IndexType offset = IndexToOffset<T, IndexType, Dims>::get(linearIndex, info) - TH_INDEX_BASE;
+    info.data[offset] = *value;
```
Comment (apaszke) [inline torch/lib/THC/THCTensorIndex.cu]: We should document that duplicate indices in `index` are UB

File: torch/lib/THC/THCTensorIndex.cu [hunk 3]
```diff
@@ -398,5 +398,53 @@ __global__ void calculateLinearIndices(
...
+template<typename IndexType, typename real, template<class, class, int> class Op, typename TensorType>
+void dispatchTakePutImpl(THCState *state, TensorType *a, TensorType *b, THCudaLongTensor *index) {
+  auto aInfo = getTensorInfo<TensorType, IndexType>(state, a);
+  aInfo.collapseDims();
+  if (aInfo.isContiguous()) {
+    auto op = Op<real, IndexType, -2>(aInfo);
+    THC_pointwiseApply2(state, b, index, op);
+  } else {
+    auto op = Op<real, IndexType, -1>(aInfo);
+    THC_pointwiseApply2(state, b, index, op);
```
Comment (apaszke) [inline torch/lib/THC/THCTensorIndex.cu]: You forgot the other 62 template instantiations :blush:

File: torch/lib/TH/generic/THTensorMath.c [hunk 4]
```diff
@@ -349,6 +349,81 @@ void THTensor_(indexCopy)(THTensor *tensor, int dim, THLongTensor *index, THTens
...
+  if (THTensor_(isContiguous)(tensor)) {
+    TH_TENSOR_APPLY2(int64_t, index, real, src,
+      int64_t linearIndex = THTensor_(wrapLinearIndex)(*index_data, numel);
+      data[linearIndex] = *src_data;
+    );
+  } else {
+    TH_TENSOR_APPLY2(int64_t, index, real, src,
+      int64_t linearIndex = THTensor_(wrapLinearIndex)(*index_data, numel);
+      int64_t dataOffset = THTensor_(dataOffset)(tensor, linearIndex);
+      data[dataOffset] = *src_data;
```
Comment (apaszke) [inline torch/lib/TH/generic/THTensorMath.c]: This is missing out-of-bound checks

File: torch/lib/TH/generic/THTensorMath.c [hunk 5]
```diff
@@ -349,6 +349,81 @@ void THTensor_(indexCopy)(THTensor *tensor, int dim, THLongTensor *index, THTens
...
+    for (ptrdiff_t i = 0; i < nIndices; i++) {
+      int64_t linearIndex = THTensor_(wrapLinearIndex)(index_data[i], srcElements);
+      dst_data[i] = src_data[linearIndex];
+    }
+  } else {
+    #pragma omp parallel for if(nIndices > TH_OMP_OVERHEAD_THRESHOLD)
+    for (ptrdiff_t i = 0; i < nIndices; i++) {
+      int64_t linearIndex = THTensor_(wrapLinearIndex)(index_data[i], srcElements);
+      int64_t dataOffset = THTensor_(dataOffset)(src, linearIndex);
+      dst_data[i] = src_data[dataOffset];
```
Comment (apaszke) [inline torch/lib/TH/generic/THTensorMath.c]: bound checks

File: torch/lib/THC/THCTensorIndex.cu (hunk 2 above)
Comment (apaszke) [reply torch/lib/THC/THCTensorIndex.cu]: \+ bound checks

File: torch/lib/THC/generic/THCTensorIndex.cu [hunk 6]
```diff
@@ -130,6 +130,37 @@ void THCTensor_(indexCopy)(THCState *state, THCTensor *dst, int dim, THCudaLongT
...
+  THCAssertSameGPU(THCudaLongTensor_checkGPU(state, 1, index));
+
+  THArgCheck(THCTensor_(nDimension)(state, dst) <= MAX_CUTORCH_DIMS, 2, CUTORCH_DIM_WARNING);
+  THArgCheck(THCTensor_(nDimension)(state, src) <= MAX_CUTORCH_DIMS, 2, CUTORCH_DIM_WARNING);
+  THArgCheck(THCudaLongTensor_nDimension(state, index) <= MAX_CUTORCH_DIMS, 2, CUTORCH_DIM_WARNING);
+
+  ptrdiff_t numIndices = THCudaLongTensor_nElement(state, index);
+
+  int srcDims = THCTensor_(nDimension)(state, src);
+  THArgCheck(srcDims > 0, 2, "Source tensor is empty");
```
Comment (apaszke) [inline torch/lib/THC/generic/THCTensorIndex.cu]: Shouldn't you check that numel of src and index match?

**Review by apaszke — COMMENTED** [review]
Looks good in general. A few minor comments. BTW `putScalar` might be useful as well

File: torch/lib/TH/generic/THTensorMath.c (hunk 1 above)
Comment (apaszke) [reply torch/lib/TH/generic/THTensorMath.c]: I thought a bit about this and I'm unsure whether we should change it or not. Consider a call like `torch.add(x, 1, out=y[0])`. Right now, if `y[0]` has a correct size, it will be filled with `x + 1`, otherwise it will be a temporary tensor that will get resized and will no longer share storage with `y`. We should do sth about it...

Comment (fmassa) [reply torch/lib/TH/generic/THTensorMath.c]: I agree it would be good to handle this. I found an example with zero-strided output tensors that is another example of something that has bugged me in the past https://github.com/torch/torch7/issues/289

Comment (ezyang) [reply torch/lib/TH/generic/THTensorMath.c]: @apaszke This feels related to #3313. Basically, if we write data into a tensor as an out, this should count as an "inplace" operation (even though its not actually inplace).

**Comment by colesbury** [comment]
I've added an accumulate option to `put_` which handles duplicate elements (by summing them). This is necessary for the backwards of `take` to be correct when indices contains duplicate elements.

(The accumulate=True option is substantially slower, so it's accumulate is False by default).

**Comment by nlgranger** [comment]
@colesbury thanks for this work and especially the accumulate option. Are thee any plan to add an axis arguent to put_ as well? Specifying the index for each individual element of a tensor can take quite a lot of place in memory.

**Comment by Alok** [comment]
It'd be nice to have an `axis` argument added, since `sklearn.utils.safe_indexing` (used by `sklearn.utils.shuffle`, which is handy for sampling multiple tensors along the same indices)   uses `X.take(indices,axis=0)` and PyTorch fails.

**Comment by fmassa** [comment]
From what I see, [numpy doesn't have an `axis` argument for `put`](https://docs.scipy.org/doc/numpy/reference/generated/numpy.put.html), and [`take` behavior can be simulated with advanced indexing](https://docs.scipy.org/doc/numpy/reference/generated/numpy.take.html).
@Alok if I understand correctly, what you are looking for can be achieved with [`tensor.index_select`](http://pytorch.org/docs/0.3.1/torch.html?highlight=index_select#torch.index_select)
